lib/src/lightclient.rs

```diff
@@ -87,28 +87,105 @@ pub struct LightClient {
 
 use serde_json::Value;
 
-fn repr_price_as_f64(from_gemini: &Value) -> f64 {
-    from_gemini
-        .get("price")
-        .unwrap()
-        .as_str()
-        .unwrap()
-        .parse::<f64>()
-        .unwrap()
+enum PriceFetchError {
+    ReqwestError(String),
+    NotJson,
+    NoElements,
+    PriceReprError(PriceReprError),
+    NanValue,
+}
+impl PriceFetchError {
+    pub(crate) fn to_string(&self) -> String {
+        use PriceFetchError::*;
+        match &*self {
+            ReqwestError(e) => format!("ReqwestError: {}", e),
+            NotJson => "NotJson".to_string(),
+            NoElements => "NoElements".to_string(),
+            PriceReprError(e) => format!("PriceReprError: {}", e),
+            NanValue => "NanValue".to_string(),
+        }
+    }
+}
+impl std::fmt::Display for PriceFetchError {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        fmt.write_str(self.to_string().as_str())
+    }
+}
+enum PriceReprError {
+    NoValue,
+    NoAsStrValue,
+    NotParseable,
+}
+impl PriceReprError {
+    fn to_string(&self) -> String {
+        use PriceReprError::*;
+        match &*self {
+            NoValue => "NoValue".to_string(),
+            NoAsStrValue => "NoAsStrValue".to_string(),
+            NotParseable => "NotParseable".to_string(),
+        }
+    }
+}
+impl std::fmt::Display for PriceReprError {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        fmt.write_str(self.to_string().as_str())
+    }
+}
+fn repr_price_as_f64(from_gemini: &Value) -> Result<f64, PriceReprError> {
+    if let Some(value) = from_gemini.get("price") {
+        if let Some(stringable) = value.as_str() {
+            if let Ok(parsed) = stringable.parse::<f64>() {
+                Ok(parsed)
+            } else {
+                Err(PriceReprError::NotParseable)
+            }
+        } else {
+            Err(PriceReprError::NoAsStrValue)
+        }
+    } else {
+        Err(PriceReprError::NoValue)
+    }
 }
 
-async fn get_recent_median_price_from_gemini() -> Result<f64, reqwest::Error> {
-    let mut trades: Vec<f64> =
-        reqwest::get("https://api.gemini.com/v1/trades/zecusd?limit_trades=11")
-            .await?
-            .json::<Value>()
-            .await?
-            .as_array()
-            .unwrap()
-            .into_iter()
-            .map(repr_price_as_f64)
-            .collect();
-    trades.sort_by(|a, b| a.partial_cmp(b).unwrap());
+async fn get_recent_median_price_from_gemini() -> Result<f64, PriceFetchError> {
```

Comment:
maybe this is more readable but could we drop all the 'let' by nesting the options/results e.g.

   150 async fn get_recent_median_price_from_gemini() -> Result<f64, PriceFetchError> {
   151     match reqwest::get("https://api.gemini.com/v1/trades/zecusd?limit_trades=11").await {
   152         Ok(httpget) => match httpget.json::<Value>().await {
   153             Ok(serialized) => {
   154                 if let Some(elements) = serialized.as_array() {
●  155                     match elements
   156                         .into_iter()
   157                         .map(repr_price_as_f64)
●  158                         .collect::<Vec<f64>>()
   159                     {
●  160                         Ok(trades) => {
   161                             if trades.iter().any(|x| x.is_nan()) {
   162                                 return Err(PriceFetchError::NanValue);
   163                             }
   164                             // NOTE:  This code will panic if a value is received that:
   165                             // 1. was parsed from a string to an f64
   166                             // 2. is not a NaN
   167                             // 3. cannot be compared to an f64
   168                             // TODO:  Show that this is impossible, or write code to handle
   169                             // that case.
   170                             trades.sort_by(|a, b| {
   171                                 a.partial_cmp(b).expect(
   172                                     "a and b are non-nan f64, I think that makes them comparable",
   173                                 )
   174                             });
   175                             Ok(trades[5])
   176                         }
●  177                         Err(e) => {
   178                             return Err(PriceFetchError::PriceReprError(e));
   179                         }
   180                     }
   181                 } else {
   182                     return Err(PriceFetchError::NoElements);
   183                 }
   184             }
   185             Err(_) => {
   186                 return Err(PriceFetchError::NotJson);
   187             }
   188         },
   189         Err(e) => {
   190             return Err(PriceFetchError::ReqwestError(e.to_string()));
   191         }
   192     }
   193 }

although this is not correct due to issues with the turbofish format on collect()

Reply:
Could do..   maybe just push your version as a subsequent PR?